src/main/java/appeng/parts/p2p/ItemP2PTunnelPart.java

```diff
@@ -120,34 +122,37 @@ public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate
                 for (ItemP2PTunnelPart target : ItemP2PTunnelPart.this.getOutputs()) {
                     final IItemHandler output = target.getAttachedItemHandler();
                     final int toSend = amountPerOutput + overflow;
-                    final ItemStack fillWithItemStack = stack.copy();
-                    fillWithItemStack.setCount(toSend);
 
-                    ItemStack received = ItemStack.EMPTY;
-
-                    for (int i = 0; i < output.getSlots(); i++) {
-                        received = output.insertItem(i, fillWithItemStack, simulate);
-
-                        if (received.isEmpty()) {
-                            break;
-                        }
+                    if (toSend <= 0) {
+                        // Both overflow and amountPerOutput are 0, so they will be for further outputs as well.
+                        break;
                     }
 
-                    overflow = received.getCount();
-                    remainder.setCount(remainder.getCount() - toSend - received.getCount());
+                    // So the documentation says that copying the stack should not be necessary because it is not
+                    // supposed to be stored or modifed by insertItem. However, ItemStackHandler will gladly store the
+                    // stack so we need to do a defensive copy. Forgecord says this is the intended behavior, and the
+                    // documentation is wrong.
+                    ItemStack stackCopy = stack.copy();
+                    stackCopy.setCount(toSend);
+                    final int sent = toSend - ItemHandlerHelper.insertItem(output, stackCopy, simulate).getCount();
 
-                    if (remainder.isEmpty()) {
-                        break;
-                    }
+                    overflow = toSend - sent;
+                    remainder -= sent;
                 }
 
                 if (!simulate) {
-                    ItemP2PTunnelPart.this.queueTunnelDrain(PowerUnits.RF, stack.getCount() - remainder.getCount());
+                    ItemP2PTunnelPart.this.queueTunnelDrain(PowerUnits.RF, amount - remainder);
                 }
             } catch (GridAccessException ignored) {
             }
 
-            return remainder;
+            if (remainder == stack.getCount()) {
+                return stack;
+            } else {
+                ItemStack copy = stack.copy();
```

Comment:
A special case for an empty itemstack might be an idea. Copying is expensive, and setting a normal itemstack to 0 also maintains the item id and so on and can have interesting side effects.

Reply:
Good point!